File: apps/rhino-cli-go/cmd/env_validate.go
package cmd

import (
	"fmt"

	"github.com/spf13/cobra"
	"github.com/wahidyankf/ose-public/apps/rhino-cli/internal/envvalidate"
)

var envValidateCmd = &cobra.Command{
	Use:   "validate",
	Short: "Validate .env.example declarations against source reads",
	Long: `Compare each app's declared env vars (infra/dev/<app>/.env.example)
against what its source code actually reads, using line-oriented regex
extractors per language. Exits non-zero when drift is detected.`,
	SilenceErrors: true,
	SilenceUsage:  false,
	RunE:          runEnvValidate,
}

func init() {
	envCmd.AddCommand(envValidateCmd)
}

func runEnvValidate(cmd *cobra.Command, _ []string) error {
	repoRoot, err := findGitRoot()
	if err != nil {
		return fmt.Errorf("failed to find git repository root: %w", err)
	}

	outputFmt, _ := cmd.Flags().GetString("output")

	var surfaces []*envvalidate.SurfaceResult
	for _, surface := range envvalidate.Surfaces {
		sr, validateErr := envvalidate.ValidateSurface(repoRoot, surface)
		if validateErr != nil {
			return fmt.Errorf("env validate failed for %s: %w", surface.App, validateErr)
		}
		surfaces = append(surfaces, sr)
	}

	result := &envvalidate.ValidateResult{Surfaces: surfaces}

	var out string
	if outputFmt == "json" {
		out, err = envvalidate.FormatJSON(result)
		if err != nil {
			return fmt.Errorf("failed to format JSON: %w", err)
		}
	} else {
		out = envvalidate.FormatText(result)
	}

	if _, printErr := fmt.Fprint(cmd.OutOrStdout(), out); printErr != nil {
		return fmt.Errorf("failed to write output: %w", printErr)
	}

	if !result.IsOK() {
		return fmt.Errorf("env validate found violations")
	}
	return nil
}

File: apps/rhino-cli-rust/src/commands/env_validate.rs
//! `env validate` command: detect drift between .env.example and source reads.
//!
//! Byte-for-byte port target: `apps/rhino-cli-go/cmd/env_validate.go`.

use anyhow::{Context, Error};
use clap::Args;

use crate::internal::cliout::OutputFormat;
use crate::internal::envvalidate::validator::validate_surface;
use crate::internal::envvalidate::{SURFACES, ValidateResult, format_json, format_text};
use crate::internal::git::root::find_root;

pub const ENV_VALIDATE_USAGE: &str = "Usage:\n  \
rhino-cli env validate [flags]\n\n\
Flags:\n  \
-h, --help   help for validate\n\n\
Global Flags:\n      \
--no-color        disable colored output\n  \
-o, --output string   output format: text, json, markdown (default \"text\")\n  \
-q, --quiet           quiet mode (errors only)\n      \
--say string      echo a message to stdout\n  \
-v, --verbose         verbose output with timestamps\n\n";

#[derive(Args, Debug)]
pub struct EnvValidateArgs {}

pub fn run_env_validate(
    _args: &EnvValidateArgs,
    output: OutputFormat,
    _verbose: bool,
    _quiet: bool,
) -> Result<(), Error> {
    let repo_root = find_root().context("failed to find git repository root")?;

    let mut surfaces = Vec::new();
    for surface in SURFACES {
        let result = validate_surface(&repo_root, surface)
            .with_context(|| format!("env validate failed for {}", surface.app))?;
        surfaces.push(result);
    }

    let result = ValidateResult { surfaces };

    let out = match output {
        OutputFormat::Json => format_json(&result).context("failed to format JSON")?,
        _ => format_text(&result),
    };

    print!("{out}");

    if !result.is_ok() {
        return Err(anyhow::anyhow!("env validate found violations"));
    }
    Ok(())
}
